### ISSUE TITLE
Infinite predicting

### ISSUE THREAD

**Issue by SimonErm** [KacperKubara/heterogeneous_dist_metrics#13]
Hello,

first i want to thank you for this great library.

I try to use the HEOM implementation with sklearns KNeighborsClassifier on a dataset with 3 numerical and 2 categorical features and about 100000 instances.

The model fits without any errors, but when i try to predict, it keep's running for hours without finishing.
When i predict without HEOM it finishes within 20 minutes.

Is the computing of HEOM that heavy or is this a bug?

**Comment by KacperKubara** [comment]
Hi @SimonErm,
it's somewhat interesting issue that is rather caused by Scikit-learn. I think I've addressed this somewhere in the issues before, but I couldn't find it so I will write it once again.

KNN is a lazy algorithm. It will store in some way all datapoints on the .fit() function and it depends on the algorithm that you use (e.g. KDTree, or BallTree).  Unfortunately, if we pass a custom metric to KNeighboursClassifier, the Scikit-Learn changes the search method to brute force. It will then have effectively computation time of O(n^2) which makes this inapplicable to use with bigger datasets. From what I could find, custom metrics are not supported with the BallTree and KDTree which have a lower complexity than the brute force search. 

When I was writing up Distython, I've managed to find and convert a BallTree implementation in Numba (this was the original code that I modified: https://gist.github.com/jakevdp/5216193) to work with my metric. The complexity then was around O(n*log(n)) which was better. Unfortunately, I don't think I can share the final code because of some contract obligations that I have from that work. This took me a fair amount time to do, because Numba is slightly difficult to use but if you have time and motivation you could try this option.

**Comment by SimonErm** [comment]
Thanks for the fast reply @KacperKubara .
I feared something like this. 
Thank you for the hint with gist. I will take a look at this.
